rfcs/text/0000-dss-deletion-daemon.md

```diff
@@ -0,0 +1,268 @@
+
+## Detailed Design
+
+*Explain the design in sufficient detail such that the implementation and (if appropriate) the interaction with existing DCP software are both reasonably clear.*
+
+### Reasons
+
+Reasons for deletion of (meta)data are
+
+* consent to share (meta)data derived from a donor's specimens is withdrawn by
+  that donor after the (meta)data was submitted to DCP
+  
+* that consent was never given but this fact is revealed only after the
+  (meta)data was submitted to the DCP
+  
+* storing or processing the (meta)data causes a serious disruption of service
+  within the DCP
+
+* the (meta)data was submitted to the DCP by a malicious party in violation of
+  the DCP's data sharing agreement. This includes (meta)data that violates the
+  laws or regulations in a jurisdiction applicable to the DCP or the (meta)data
+  it stores
+
+Whether any of the reasons require a physical or logical deletion is at the
+discretion of the administrator performing the deletion. That discretion is
+guided by a standard operating procedure (SOP).
+
+### Process
+
+Once it is determined that a (meta)data file needs to be physically deleted –
+the details of this determination are the scope of an SOP – the following
+procedure takes place:
+
+1) An administrator identifies the Data Store file versions to be deleted. The
+   administrator then identifies the Data Store bundle versions referencing
+   those file versions.
+
+2) The administrator runs the admin CLI tool to queue the bundle for deletion using the bundles UUID and version.
+   The bundle UUID and version is required for deletion. The admin CLI will return a list of files that will be 
+   deleted and bundles that are effected by this deletion. Bundles may be affected if they share files or are derived 
+   from the deleted bundle.
+   
+   ```bash
+   delete_bundle.py <uuid>.<version> --reason [consent_withdrawn, consent_absent, service_disruption, legal] 
+    --type [logical, physical] --details "<additional info>"
+   
+   Files to be deleted:
+   <fileblobs>
+   .
+   .
+   .
+
+   Effected Bundles:
+   <uuid>.<version>
+   .
+   .
+   .
+
+   confirm (yes or no)?
+
+   ```
+
+3) **Logical Deletion:** The DSS places a **bundle tombstone** in the
+   underlying storage (S3 or GCP bucket). The tombstone's content is as follows:
+   ```JSON
+   { "reason": "<reason>",
+     "details": "<additional info>",
+     "admin_email": "<admin>"
+   }
+   ```
+   The tombstone causes `GET /bundle/{uuid}` to return a 404. If `{version}` 
+   signifies the latest version of the bundle
+   identified by `{uuid}`, requests for `GET /bundle/{uuid}` without a version
+   also return a 404. The same applies to HEAD requests, both versioned and
+   unversioned. The file versions referenced by the deleted bundle version are
+   not immediately affected by the deletion and remain accessible. The bundle and associated files are added to 
+   a deletion queue for regularly schedule **Physical Deletion**.
```

Comment:
We should discuss physical deletion being continent on specific reasons of a logical ("soft") delete. See https://landing.google.com/sre/sre-book/chapters/data-integrity/#how-google-sre-faces-the-challenges-of-data-integrity-G1sYCe